packages/xo-web/src/xo-app/user/otp.js

```diff
@@ -0,0 +1,59 @@
+import authenticator from 'otplib/authenticator'
+import crypto from 'crypto'
+import decorate from 'apply-decorators'
+import qrcode from 'qrcode'
+import React from 'react'
+import { addOtp, removeOtp } from 'xo'
+import { injectState, provideState } from 'reaclette'
+
+import { Container, Row, Col } from '../../common/grid'
+import { Toggle } from '../../common/form'
+
+authenticator.options = { crypto }
+
+export default decorate([
+  provideState({
+    initialState: ({ user }) => ({
+      secret: user.preferences.otp || authenticator.generateSecret(),
+    }),
+    effects: {
+      _handleOtp: (_, isChecked) => ({ secret }, _) => {
+        if (isChecked) {
+          addOtp(secret)
+        } else {
+          removeOtp()
+        }
```

Comment:
[Effects](https://github.com/JsCommunity/reaclette#effects--string-effect-) can return a reducer (4th point) to mutate the state.

Because you don't do that here it would be better to write this:

```js
_handleOtp (_, isChecked) {
  return isChecked ? addOtp(this.state.secret) : remoteOtp()
}
```